src/ReplicatedStorage/Client/InterfaceComponents/Toolbar/ToolbarItem.lua

```diff
@@ -0,0 +1,234 @@
+
+	-- Sizing states
+	local inputNameTextBounds = Value(Vector2.new())
+	local toolNameTextBounds = Value(Vector2.new())
+	local absoluteButtonSize = Value(Vector2.new())
+
+
+	-- Size States
+	local paddingSize = Computed(function()
+		local currentAbsoluteButtonSize = absoluteButtonSize:get()
+
+		if currentAbsoluteButtonSize == nil then
+			return 0
+		else
+			return currentAbsoluteButtonSize.Y / 10
+		end
+	end)
+
+	local buttonSize = Spring(Computed(function()
+		local isVisible = visible:get()
+		local currentAbsoluteButtonSize = absoluteButtonSize:get()
+		local currentInputNameTextBounds = inputNameTextBounds:get()
+		local currentToolNameTextBounds = toolNameTextBounds:get()
+
+		if currentAbsoluteButtonSize == nil or currentInputNameTextBounds == nil or currentToolNameTextBounds == nil or isVisible == false then
+			return UDim2.fromScale(0, 1)
+		else
+			local width = paddingSize:get() * 4 + currentInputNameTextBounds.X + currentToolNameTextBounds.X
+			return UDim2.new(0, width, 1, 0)
+		end
+	end), 20, 0.9)
+
+	local buttonContainerSize = Spring(Computed(function()
+		local isVisible = visible:get()
+		local currentAbsoluteButtonSize = absoluteButtonSize:get()
+
+		if not isVisible or currentAbsoluteButtonSize == nil then
+			return UDim2.fromScale(0, 1)
+		else
+			return UDim2.new(0, absoluteButtonSize:get().X, 1, 0) + UDim2.fromOffset(paddingSize:get() * 2, 0)
+		end
+	end), 20, 0.9)
+
+	local textContainerSize = Computed(function()
+		local currentInputNameTextBounds = inputNameTextBounds:get()
+		local currentToolNameTextBounds = toolNameTextBounds:get()
+
+		if currentInputNameTextBounds == nil or currentToolNameTextBounds == nil then
+			return UDim2.fromScale(0, 1)
+		else
+			local width = currentInputNameTextBounds.X + currentToolNameTextBounds.X + paddingSize:get() * 3
+			return UDim2.new(0, width, 1, 0)
+		end
+	end)
+
+
+	-- Position States
+	local inputNameTextPosition = Computed(function()
+		return UDim2.new(0, paddingSize:get(), 0.5, 0)
+	end)
+
+	local toolNameTextPosition = Computed(function()
+		local currentInputNameTextBounds = inputNameTextBounds:get()
+
+		if currentInputNameTextBounds == nil then
+			return UDim2.new()
+		else
+			return UDim2.new(0, currentInputNameTextBounds.X + paddingSize:get() * 2, 0.5, 0)
+		end
+	end)
+
+
+	-- Transparency/Thickness States
+	local backgroundTransparency = Spring(Computed(function()
+		if selected:get() and hovering:get() then
+			return 0.2
+		elseif selected:get() then
+			return 0.3
+		elseif hovering:get() then
+			return 0.4
+		else
+			return 0.525
+		end
+	end), 20, 0.9)
+
+	local strokeTransparency = Spring(Computed(function()
+		return if selected:get() then 0 else 1
+	end), 20, 0.9)
+
+	local strokeThickness = Spring(Computed(function()
+		return if selected:get() then SharedInterfaceState.StrokeSize:get() else 0
+	end), 20, 0.9)
+
+	local textTransparency = Spring(Computed(function()
+		return if visible:get() then 0 else 1
+	end), 20, 0.9)
```

Comment:
There are a lot of magic numbers in this code - perhaps they should be abstracted to a config file.